File: daily_problems/2024/09/0909/personal_submission/cf551b_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include <bits/extc++.h>
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    string a, b, c;
    cin >> a >> b >> c;

    vector<int> cnt_a(26, 0), cnt_b(26, 0), cnt_c(26, 0);
    for (auto ch: a) cnt_a[ch - 'a'] ++;
    for (auto ch: b) cnt_b[ch - 'a'] ++;
    for (auto ch: c) cnt_c[ch - 'a'] ++;

    int x = 0, y = 0;
    for (int i = 0; i * b.size() < a.size(); i ++) {
        bool flg = true;
        int cnt = a.size();
        for (int j = 0; j < 26; j ++) {
            if (cnt_a[j] < cnt_b[j] * i) {
                flg = false;
                break;
            }
            if (cnt_c[j]) cnt = min(cnt, (cnt_a[j] - cnt_b[j] * i) / cnt_c[j]);
        }
        if (!flg) break;
        if (i + cnt > x + y) x = i, y = cnt;
    }
    for (int i = 0; i < x; i ++) cout << b;
    for (int i = 0; i < y; i ++) cout << c;
    for (int i = 0; i < 26; i ++)  cout << string(cnt_a[i] - cnt_b[i] * x - cnt_c[i] * y, 'a' + i);

    return 0;
}

File: daily_problems/2024/09/0909/personal_submission/cf551b_yawn_sean.py
# Submission link: https://codeforces.com/contest/551/submission/280381501
def main():
    a = I()
    b = I()
    c = I()

    cnt_a = [0] * 26
    cnt_b = [0] * 26
    cnt_c = [0] * 26

    for x in a: cnt_a[ord(x) - ord('a')] += 1
    for x in b: cnt_b[ord(x) - ord('a')] += 1
    for x in c: cnt_c[ord(x) - ord('a')] += 1

    x, y = 0, 0

    for i in range(len(a) // len(b) + 1):
        cnt = len(a)
        flg = True
        for j in range(26):
            if cnt_a[j] < cnt_b[j] * i:
                flg = False
            if cnt_c[j]:
                cnt = fmin(cnt, (cnt_a[j] - cnt_b[j] * i) // cnt_c[j])
        if not flg: break
        if i + cnt > x + y:
            x, y = i, cnt

    outs = []
    outs.append(b * x)
    outs.append(c * y)

    for i in range(26):
        outs.append((cnt_a[i] - cnt_b[i] * x - cnt_c[i] * y) * chr(ord('a') + i))

    print(''.join(outs))